// @sentry/react-native/dist/js/integrations/expoupdateslistener.js
import { addBreadcrumb, debug } from '@sentry/core';
import { isExpo, isExpoGo } from '../utils/environment';
const INTEGRATION_NAME = 'ExpoUpdatesListener';
const BREADCRUMB_CATEGORY = 'expo.updates';
/**
 * Tries to load `expo-updates` and retrieve exports needed by this integration.
 * Returns `undefined` if `expo-updates` is not installed.
 */
function getExpoUpdatesExports() {
    try {
        const expoUpdates = require('expo-updates');
        if (typeof expoUpdates.addUpdatesStateChangeListener === 'function') {
            return expoUpdates;
        }
    }
    catch (_) {
        // that happens when expo-updates is not installed
    }
    return undefined;
}
const STATE_TRANSITIONS = [
    { field: 'isChecking', message: 'Checking for update', level: 'info' },
    {
        field: 'isUpdateAvailable',
        message: 'Update available',
        level: 'info',
        getData: ctx => {
            var _a;
            const updateId = (_a = ctx.latestManifest) === null || _a === void 0 ? void 0 : _a.id;
            return updateId ? { updateId } : undefined;
        },
    },
    { field: 'isDownloading', message: 'Downloading update', level: 'info' },
    {
        field: 'isUpdatePending',
        message: 'Update downloaded',
        level: 'info',
        getData: ctx => {
            var _a;
            const updateId = (_a = ctx.downloadedManifest) === null || _a === void 0 ? void 0 : _a.id;
            return updateId ? { updateId } : undefined;
        },
    },
    {
        field: 'checkError',
        message: 'Update check failed',
        level: 'error',
        getData: ctx => ({
            error: ctx.checkError.message || String(ctx.checkError),
        }),
    },
    {
        field: 'downloadError',
        message: 'Update download failed',
        level: 'error',
        getData: ctx => ({
            error: ctx.downloadError.message || String(ctx.downloadError),
        }),
    },
    {
        field: 'rollback',
        message: 'Rollback directive received',
        level: 'warning',
        getData: ctx => ({
            commitTime: ctx.rollback.commitTime,
        }),
    },
    { field: 'isRestarting', message: 'Restarting for update', level: 'info' },
];
/**
 * Listens to Expo Updates native state machine changes and records
 * breadcrumbs for meaningful transitions such as checking for updates,
 * downloading updates, errors, rollbacks, and restarts.
 */
export const expoUpdatesListenerIntegration = () => {
    let subscription;
    function setup(client) {
        client.on('afterInit', () => {
            var _a;
            if (!isExpo() || isExpoGo()) {
                return;
            }
            const expoUpdates = getExpoUpdatesExports();
            if (!expoUpdates) {
                debug.log('[ExpoUpdatesListener] expo-updates is not available, skipping.');
                return;
            }
            // Remove any previous subscription to prevent duplicate breadcrumbs
            // if Sentry.init() is called multiple times.
            subscription === null || subscription === void 0 ? void 0 : subscription.remove();
            // Seed with the current state so that the first event does not
            // generate spurious breadcrumbs for already-truthy fields.
            let previousContext = (_a = expoUpdates.latestContext) !== null && _a !== void 0 ? _a : {};
            subscription = expoUpdates.addUpdatesStateChangeListener((event) => {
                const ctx = event.context;
                handleStateChange(previousContext, ctx);
                previousContext = ctx;
            });
        });
        client.on('close', () => {
            subscription === null || subscription === void 0 ? void 0 : subscription.remove();
            subscription = undefined;
        });
    }
    return {
        name: INTEGRATION_NAME,
        setup,
    };
};
/**
 * Compares previous and current state machine contexts and emits
 * breadcrumbs for meaningful transitions (falsy→truthy).
 *
 * @internal Exposed for testing purposes
 */
export function handleStateChange(previous, current) {
    var _a;
    for (const transition of STATE_TRANSITIONS) {
        if (!previous[transition.field] && current[transition.field]) {
            addBreadcrumb({
                category: BREADCRUMB_CATEGORY,
                message: transition.message,
                level: transition.level,
                data: (_a = transition.getData) === null || _a === void 0 ? void 0 : _a.call(transition, current),
            });
        }
    }
}
